### PR TITLE
fix(terminal): work around Neovim <0.12.2 terminal paste fragmentation

### LINKED ISSUE #161: [BUG]
## Bug Description

Pasting text with Cmd+V (macOS) in the Claude Code terminal window truncates content, while right-click paste works correctly. The terminal shows the correct number of pasted lines (e.g.,
"[Pasted text #1 +109 lines]") with right-click paste, but Cmd+V only pastes a partial amount of the content.

## To Reproduce

1. Open Neovim with the claudecode.nvim plugin configured
2. Start Claude Code terminal with `<leader>ac` (or `:ClaudeCode`)
3. Copy a large amount of text (e.g., 100+ lines) to system clipboard
4. In the Claude Code terminal, press Cmd+V to paste
5. Observe that only a portion of the text is pasted (truncated)
6. Compare with right-click → Paste, which correctly pastes all content

## Expected Behavior

Cmd+V should paste the entire clipboard contents into the Claude Code terminal, identical to the behavior of right-click paste. The terminal should display the bracketed paste indicator showing
 all lines were pasted (e.g., "[Pasted text #N +109 lines]").

## Environment

- Neovim version: 0.11.5
- Claude Code CLI version: 2.0.5
- OS: macOS 15.7.2
- Terminal: WezTerm
- Plugin configuration:
  ```lua
  {
    "coder/claudecode.nvim",
    dependencies = { "folke/snacks.nvim" },
    opts = {
      terminal = {
        snacks_win_opts = {
          position = "float",
          width = 0.9,
          height = 0.9,
          border = "rounded",
        },
      },
    },
  }
- Neovim clipboard setting: vim.opt.clipboard = "unnamedplus"

Observed Behavior

- Right-click paste: ✅ Works correctly - shows "[Pasted text #14 +109 lines]"
- Cmd+V paste: ❌ Truncates content - only pastes partial text
- Sometimes shows a few extra characters at the beginning of Cmd+V paste

Debugging Attempted

Tried various terminal mode paste keybindings without success:
- vim.keymap.set("t", "<D-v>", function() return vim.fn.getreg("+") end, { expr = true })
- vim.keymap.set("t", "<D-v>", '<C-\\><C-n>"+pi')
- vim.api.nvim_chan_send(chan, content)
- vim.api.nvim_paste(reg, true, -1)

None of these approaches replicated the successful right-click paste behavior.

Additional Context

The issue appears to be related to how Cmd+V is processed in terminal mode. Right-click paste triggers proper bracketed paste mode in the terminal, but Cmd+V seems to use a different paste
mechanism that truncates content. This may be an interaction between:
1. Neovim's clipboard = "unnamedplus" setting
2. Terminal mode key handling
3. The snacks.nvim terminal implementation

The same Cmd+V paste works correctly in regular Neovim buffers and other terminal emulators outside the Claude Code floating window.

This provides all the details needed for the plugin maintainers to investigate and reproduce the issue. Would you like me to help you submit this to the GitHub repository?

### PR DESCRIPTION
**Pull request by ThomasK33**
Fixes #161

## Problem

Pasting a large block (Cmd+V) into the Claude Code terminal appeared to **truncate** — the terminal showed multiple `[Pasted text #N]` placeholders with stray characters, while right-click paste worked.

## Root cause (upstream Neovim, not this plugin)

This is an upstream Neovim core bug — [`neovim/neovim#39110`](https://github.com/neovim/neovim/issues/39110), fixed by [`#39152`](https://github.com/neovim/neovim/pull/39152) in **0.12.2**. On affected Neovim the default `vim.paste` runs `nvim_put()` once per streamed phase (1→2→3), and each write into a `:terminal` is independently wrapped in bracketed-paste markers (`ESC[200~ … ESC[201~`), so one logical paste reaches Claude as **N separate paste events**. The plugin never touched the paste path — it inherited the behavior. The fix was **not** backported to the 0.11 line (0.11.7 still reproduces), so users there can't fix it by upgrading in place.

## What this PR does

Adds a scoped, version-gated `vim.paste` compatibility shim (`lua/claudecode/terminal/paste_fix.lua`), controlled by a new `terminal.fix_streamed_paste` option:

- **`"auto"` (default)** — enabled only on affected Neovim (`< 0.12.2`); a no-op otherwise. `true` forces it on, `false` disables it.
- **Scoped** to the plugin's own managed terminal buffer; delegates to the existing `vim.paste` for every other buffer, mode, and `phase == -1`.
- **Coalesces** the streamed phases into one `phase == -1` replay (one bracketed segment), with a mid-line seam-join so no spurious newlines are introduced.
- **Hot-path safe**: the managed-terminal decision is latched once per stream (so a mid-stream focus change can't strand buffered content), the provider lookup is `pcall`-guarded (can't break pasting editor-wide), and the config is dynamic (a later `setup{ fix_streamed_paste = false }` disables it without uninstalling).

The real fix remains **upgrade to Neovim 0.12.2+**; this is a compatibility shim for users on 0.11.x / 0.12.0 / 0.12.1.

## Verification

Reproduced and fixed empirically with a new `fixtures/paste-repro/` harness (a bracketed-paste "observer" stands in for `claude` and counts how many `ESC[200~`/`ESC[201~` segments the inner PTY receives), driven via [`agent-tty`](https://github.com/coder/agent-tty):

| Neovim | default | with shim |
| --- | --- | --- |
| 0.11.5 / 0.11.6 / 0.11.7 / 0.12.0 / 0.12.1 | 🐛 6 segments | ✅ 1 segment |
| 0.12.2 | ✅ 1 segment | ✅ 1 segment (no-op) |

The shipped shim's output is **byte-identical** to fixed-Neovim (0.12.2) output across edge cases: newline-on-chunk-boundary, consecutive blank lines, no trailing newline, CRLF, and a 20 KB multi-boundary paste.

## Test plan

- `mise run all` → 456 tests pass, luacheck 0 warnings, treefmt clean.
- 26 new unit tests in `tests/unit/terminal/paste_fix_spec.lua` (coalescing incl. phase 2, mid-stream focus change, dynamic disable, version gate, seam-join edge cases).
- End-to-end re-verified on Neovim 0.11.7 (raw bug reproduces; shipped shim collapses 6→1) via `fixtures/paste-repro/agent-repro.sh`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by ThomasK33** [comment]
@codex review

**Comment by ThomasK33** [comment]
@claude review

File: lua/claudecode/terminal/paste_fix.lua [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+      return orig_paste(lines, phase)
+    end
+
+    M._accumulate(chunks, lines)
+
+    if phase == 3 then
+      local buffered = chunks
+      chunks = {}
+      streaming = false
+      return orig_paste(buffered, -1) -- one non-streamed replay = one bracketed segment
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply lua/claudecode/terminal/paste_fix.lua]: Good catch — fixed in 5e8f5f1.

The shim now captures the target buffer at phase 1. At phase 3, if focus is still on that terminal it replays through the original `vim.paste` as before (common case, unchanged). If focus has left, it no longer replays via the original handler (which would dump the buffered text into the now-current buffer); instead it sends the coalesced content straight to the captured terminal's channel as a single bracketed-paste segment (`ESC[200~ … ESC[201~`, lines joined with `\n` to match the bytes Neovim writes to the inner PTY). Added a unit test asserting the channel redirect.

(Note this mid-stream focus change is essentially unreachable from a TUI bracketed paste — phases 1→3 are processed as one atomic input burst — but the redirect makes the behavior correct regardless.)

**Comment by ThomasK33** [comment]
@codex review

**Comment by ThomasK33** [comment]
@claude review

File: lua/claudecode/terminal/paste_fix.lua [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+      if vim.api.nvim_get_current_buf() == target_buf then
+        return orig_paste(buffered, -1)
+      end
+      -- Focus left the terminal mid-stream: replaying via the original handler
+      -- would dump the buffered text into whatever buffer is now current. Send
+      -- it straight to the captured terminal's channel instead, matching the
+      -- bytes Neovim would have written (one bracketed-paste segment).
+      local chan = vim.api.nvim_buf_is_valid(target_buf) and vim.bo[target_buf].channel
+      if chan and chan > 0 then
+        pcall(vim.api.nvim_chan_send, chan, "\27[200~" .. table.concat(buffered, "\n") .. "\27[201~")
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply lua/claudecode/terminal/paste_fix.lua]: Good point — fixed in da05ebc by removing the raw channel write entirely.

You're right that only Neovim's terminal paste knows whether the child enabled bracketed paste, and that state isn't queryable from Lua (confirmed: `nvim_get_chan_info` exposes `mode`/`pty`/`buffer` but not the DECSET 2004 state). Rather than reconstruct the `ESC[200~…ESC[201~` framing with an assumption, the phase-3 replay now just delegates to the original `vim.paste` handler — i.e. matching `vim.paste` behavior, as you suggested. I also dropped the now-unused target-buffer capture.

Note a focus change between phases isn't actually reachable from a TUI bracketed paste (it's processed as one atomic input burst), so this path is a safety fallback; in the worst case it now matches stock Neovim's behavior instead of corrupting input.

**Comment by ThomasK33** [comment]
@codex review

**Comment by ThomasK33** [comment]
@claude review

File: lua/claudecode/terminal/paste_fix.lua [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  if type(v) ~= "table" or type(v.major) ~= "number" then
+    return false
+  end
+  if v.major ~= 0 then
+    return false -- 1.0+ is well past the fix
+  end
+  local minor = v.minor or 0
+  local patch = v.patch or 0
+  return minor < 12 or (minor == 12 and patch < 2)
+end
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply lua/claudecode/terminal/paste_fix.lua:42]: Addressed in 8a50bfe.

`is_affected_version` now treats a `0.12.2` build whose `vim.version().prerelease` is set (e.g. `"dev"`) as affected, since such a nightly predates the 0.12.2 release that carries the backport. The 0.12.2 release (`prerelease == nil`) and 0.13+ remain unaffected. Enabling the shim on a build that already has the fix is a verified no-op (coalescing produces byte-identical output), so erring toward "affected" at the boundary is safe. Added unit tests for the prerelease boundary.

**Comment by ThomasK33** [comment]
@codex review

**Comment by ThomasK33** [comment]
@claude review

File: lua/claudecode/terminal/paste_fix.lua [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    end
+
+    M._accumulate(chunks, lines)
+
+    if phase == 3 then
+      local buffered = chunks
+      chunks = {}
+      streaming = false
+      -- Replay as a single non-streamed paste through the original handler. We
+      -- deliberately delegate rather than writing to the terminal channel
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply lua/claudecode/terminal/paste_fix.lua]: Addressed in e49e2f4 — and you're right, the previous delegation didn't aim the replay at the captured terminal.

The shim now captures the terminal's **window** at phase 1. At phase 3, if focus has moved off it, the coalesced replay runs inside that window via `nvim_win_call(target_win, ...)`. This reaches the terminal even from another window, and crucially still goes through Neovim's own terminal paste — so it respects the child's bracketed-paste state (the round-2 concern) rather than reconstructing `ESC[200~/ESC[201~` ourselves (which isn't queryable from Lua).

Verified in a headless Neovim: with focus on a separate normal buffer (normal mode), `nvim_win_call(term_win, () -> vim.paste(buffered, -1))` delivers exactly one correctly-framed bracketed segment to the terminal child (`\x1b[200~AAA\nBBB\x1b[201~`) and leaves the now-current buffer untouched. The common case (focus unchanged) still replays directly with no `win_call`.

**Comment by ThomasK33** [comment]
@codex review

**Comment by ThomasK33** [comment]
@claude review
